cpp/dolfinx/fem/CoordinateElement.h

```diff
@@ -49,18 +49,47 @@ class CoordinateElement
   /// Return the topological dimension of the cell shape
   int topological_dimension() const;
 
-  /// Tabulate shape functions up to n-th order derivative at points X in the
-  /// reference geometry
-  /// Note: Dynamic allocation.
+  /// Compute basis values and derivatives at set of points.
+  /// @param[in] n The order of derivatives, up to and including, to
+  /// compute. Use 0 for the basis functions only.
+  /// @param[in] X The points at which to compute the basis functions.
+  /// The shape of x is (number of points, geometric dimension).
+  /// @return The basis functions (and derivatives). The shape is
+  /// (derivative, number point, number of basis fn, value size).
   xt::xtensor<double, 4> tabulate(int n, const array2d<double>& X) const;
 
-  /// Compute J, K and detJ for a cell with given geometry, and the
-  /// basis functions and first order derivatives at points X
-  void compute_jacobian_data(const xt::xtensor<double, 4>& tabulated_data,
-                             const array2d<double>& X,
-                             const array2d<double>& cell_geometry,
-                             std::vector<double>& J, xtl::span<double> detJ,
-                             std::vector<double>& K) const;
+  /// Compute Jacobian for a cell with given geometry using the
+  /// basis functions and first order derivatives.
+  /// @param[in] tabulated_data Pre-computed basis values and derivatives at set
+  /// of points. Need at least first order derivative.
+  /// @param[in] cell_geometry Coordinates/geometry
+  /// @param[in,out] J The Jacobian
+  /// The shape of J is (number of points, geometric dimension, topological
+  /// dimenson).
+  void compute_jacobian(const xt::xtensor<double, 4>& tabulated_data,
+                        const xt::xtensor<double, 2>& cell_geometry,
```

Comment:
Maybe allow data for multiple cells to be passed? Common operation would be evaluating data at same reference points on multiple cells. Would there be performance benefits?

Reply:
I'm not sure if there would be performance benefits.
The shape function and derivatives are computed only once per mesh (or coordinate element), and the main computation is just a dot product.